source/extensions/transport_sockets/tls/context_impl.cc

```diff
@@ -1080,25 +1084,56 @@ enum ssl_select_cert_result_t
 ServerContextImpl::selectTlsContext(const SSL_CLIENT_HELLO* ssl_client_hello) {
   const bool client_ecdsa_capable = isClientEcdsaCapable(ssl_client_hello);
   const bool client_ocsp_capable = isClientOcspCapable(ssl_client_hello);
+  std::string sni =
+      std::string(SSL_get_servername(ssl_client_hello->ssl, TLSEXT_NAMETYPE_host_name));
 
-  // Fallback on first certificate.
-  const TlsContext* selected_ctx = &tls_contexts_[0];
-  auto ocsp_staple_action = ocspStapleAction(*selected_ctx, client_ocsp_capable);
-  for (const auto& ctx : tls_contexts_) {
-    if (client_ecdsa_capable != ctx.is_ecdsa_) {
-      continue;
-    }
+  const TlsContext* selected_ctx = nullptr;
 
-    auto action = ocspStapleAction(ctx, client_ocsp_capable);
-    if (action == OcspStapleAction::Fail) {
-      continue;
-    }
+  // do SNI matching and signature algorithm matching if SNI exists
+  if (!sni.empty()) {
+    for (const auto& ctx : tls_contexts_) {
```

Comment:
This algorithm won't scale well for larger numbers of certificates. I think it would be better to copy the algorithm we have for filter chain matching for SNI, see source/server/filter_chain_manager_impl.cc, using a hash-table lookup.